### ISSUE TITLE
Errors received during "make" (post ./configure) due to libxdp.c file. Need to configure and make xdp-tools prior to compile from the root directory

### ISSUE THREAD

**Issue by hundredmiles** [xdp-project/bpf-examples#66]
After cloning the repo and running ```./configure``` from the root, if I run ```make``` from the root or from the directory of interest (e.g. - ```pping```), I get lots of error related to the file ```libxdp.c``` inspite of all the necessary files being present in the directories ```libbpf``` and ```libxdp``` that are cloned during ```./configure```. 

However, if I go to ```/lib/xdp-tools``` and run ```./configure``` followed by ```make``` and then come back to the root folder or the directory of interest and run ```make``` again, the errors are not seen.

So, in a nutshell
```
Expected but not working: clone repo --> ./configure (from root) --> make (from root)
Workaround: clone repo --> ./configure (from root) --> go to /lib/xdp-tools/ --> ./configure (from xdp-tools) --> make (from xdp-tools) --> come back to root and run make (from root)
```

Note that I have tried running  ```git submodule update --init ``` after cloning the repo as mentioned by @simosund in issue #63, but get the same errors.

Ubuntu - 20.04 LTS

Here is small snippet of the error. 

```
anindya@ubuntu:testebpf/pping$ make
  libbpf
    CC       install/lib/libbpf.a
    INSTALL  install/lib/libbpf.a

  libxdp
    CC       install/lib/libxdp.a
libxdp.c:243:20: error: static declaration of ‘btf__load_from_kernel_by_id’ follows non-static declaration
  243 | static struct btf *btf__load_from_kernel_by_id(__u32 id)
      |                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from libxdp.c:29:
/home/anindya/testebpf/lib/install/include/bpf/btf.h:121:24: note: previous declaration of ‘btf__load_from_kernel_by_id’ was here
  121 | LIBBPF_API struct btf *btf__load_from_kernel_by_id(__u32 id);
      |                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~
libxdp.c: In function ‘btf__load_from_kernel_by_id’:
libxdp.c:248:2: error: ‘btf__get_from_id’ is deprecated: libbpf v0.6+: use btf__load_from_kernel_by_id instead [-Werror=deprecated-declarations]
  248 |  err = btf__get_from_id(id, &btf);
      |  ^~~
In file included from libxdp.c:29:
/home/anindya/testebpf/lib/install/include/bpf/btf.h:124:16: note: declared here
  124 | LIBBPF_API int btf__get_from_id(__u32 id, struct btf **btf);
      |                ^~~~~~~~~~~~~~~~
libxdp.c: At top level:
libxdp.c:256:14: error: static declaration of ‘btf__type_cnt’ follows non-static declaration
  256 | static __u32 btf__type_cnt(const struct btf *btf)
      |              ^~~~~~~~~~~~~
In file included from libxdp.c:29:
/home/anindya/testebpf/lib/install/include/bpf/btf.h:137:18: note: previous declaration of ‘btf__type_cnt’ was here
  137 | LIBBPF_API __u32 btf__type_cnt(const struct btf *btf);
      |                  ^~~~~~~~~~~~~
libxdp.c: In function ‘btf__type_cnt’:
libxdp.c:259:2: error: ‘btf__get_nr_types’ is deprecated: libbpf v0.7+: use btf__type_cnt() instead; note that btf__get_nr_types() == btf__type_cnt() - 1 [-Werror=deprecated-declarations]
  259 |  return btf__get_nr_types(btf) + 1;
      |  ^~~~~~
```

Am I missing anything or this is a configuration issue in the repo?

**Comment by simosund** [comment]
I cannot reproduce this issue myself (the `Expected but not working`-steps work for me). As it seems to be related to the libxdp subproject (`/lib/xdp-tools`) I think @tohojo might stand a better shot at figuring it out. Until we get it fixed other people who may run into the same issue can hopefully find this issue and use the workaround described by @hundredmiles. 

While not necessarily urgent, I think we should keep this issue in mind. If people need to do a bunch of troubleshooting and in general go through a hassle just to get `bpf-examples` to compile it makes it much less useful as as an entry-point for people getting started with BPF. Hopefully we can get the configuration fixed so that the extra step of having to explicitly configure and compile the subprojects as well, but regardless just adding some basic instructions on how to setup `bpf-examples` might be a good idea (even if they should hopefully just be `clone -> ./configure -> make`).

**Comment by simosund** [comment]
On a somewhat related topic I can't get nat64-bpf to compile. First of it requires `libmnl` which the configuration doesn't check for, and even after installing `libmnl` I get some error (see below) compiling it which seem to hint at some type of dependency still missing. I think @hundredmiles had some issues with this as well in #63. So unless the compilation process for `nat64-bpf` can be simplified it should perhaps not be part of the programs compiled by default when just running make from the root directory.

```
nat64.c: In function ‘main’:
nat64.c:381:15: warning: implicit declaration of function ‘nat64_kern__open’ [-Wimplicit-function-declaration]
  381 |         obj = nat64_kern__open();
      |               ^~~~~~~~~~~~~~~~
nat64.c:381:13: warning: assignment to ‘struct nat64_kern *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
  381 |         obj = nat64_kern__open();
      |             ^
nat64.c:391:12: error: invalid use of undefined type ‘struct nat64_kern’
  391 |         obj->bss->config = cfg.c;
      |            ^~
nat64.c:392:37: error: invalid use of undefined type ‘struct nat64_kern’
  392 |         bpf_map__set_max_entries(obj->maps.v6_state_map, num_addr);
      |                                     ^~
nat64.c:393:37: error: invalid use of undefined type ‘struct nat64_kern’
  393 |         bpf_map__set_max_entries(obj->maps.v4_reversemap, num_addr);
      |                                     ^~
nat64.c:394:37: error: invalid use of undefined type ‘struct nat64_kern’
  394 |         bpf_map__set_max_entries(obj->maps.reclaimed_addrs, num_addr);
      |                                     ^~
nat64.c:396:15: warning: implicit declaration of function ‘nat64_kern__load’ [-Wimplicit-function-declaration]
  396 |         err = nat64_kern__load(obj);
      |               ^~~~~~~~~~~~~~~~
nat64.c:408:58: error: invalid use of undefined type ‘struct nat64_kern’
  408 |                 err = bpf_map_update_elem(bpf_map__fd(obj->maps.allowed_v6_src),
      |                                                          ^~
nat64.c:417:53: error: invalid use of undefined type ‘struct nat64_kern’
  417 |         attach_ingress.prog_fd = bpf_program__fd(obj->progs.nat64_ingress);
      |                                                     ^~
nat64.c:424:52: error: invalid use of undefined type ‘struct nat64_kern’
  424 |         attach_egress.prog_fd = bpf_program__fd(obj->progs.nat64_egress);
      |                                                    ^~
nat64.c:460:9: warning: implicit declaration of function ‘nat64_kern__destroy’ [-Wimplicit-function-declaration]
  460 |         nat64_kern__destroy(obj);
      |         ^~~~~~~~~~~~~~~~~~~
make[1]: *** [../lib/common.mk:83: nat64] Error 1
make: *** [Makefile:37: nat64-bpf] Error 2
```

**Comment by tohojo** [comment]
Could you please both test #67 ?

**Comment by simosund** [comment]
So I was never able to replicate the issue @hundredmiles has in the first place, so all I can report is that just running `./configure -> make` still works for me (except I still get the same error with `nat64-bpf` posted above).

**Comment by tohojo** [comment]
Simon Sundberg ***@***.***> writes:

> So I was never able to replicate the issue @hundredmiles has in the
> first place, so all I can report is that just running `./configure ->
> make` still works for me (except I still get the same error with
> `nat64-bpf` posted above).

Could you please try doing a 'make clean' inside the nat64-bpf
directory, then running 'make' again and posting the full output?

**Comment by hundredmiles** [comment]
I am still getting an error
Here is the output of  ```./configure ``` followed by ```make```

```
anindya@ubuntu:~/99/bpf-examples$ CLANG=clang-14 LLC=llc-14 ./configure 
clang: 14.0.6-++20221029082011+f28c006a5895-1~exp1~20221029202041.162
libbpf support: submodule
Submodule 'lib/libbpf' (https://github.com/xdp-project/libbpf.git) registered for path 'lib/libbpf'
Submodule 'lib/xdp-tools' (https://github.com/xdp-project/xdp-tools) registered for path 'lib/xdp-tools'
Cloning into '/home/anindya/99/bpf-examples/lib/libbpf'...
Cloning into '/home/anindya/99/bpf-examples/lib/xdp-tools'...
Submodule path 'lib/libbpf': checked out '2cd2d03f63242c048a896179398c68d2dbefe3d6'
Submodule path 'lib/xdp-tools': checked out 'c6befe331d07ab35a89f093ca68c7aa662fb6b7d'
ELF support: yes
zlib support: yes
libxdp support: submodule
Configuring libxdp to use our libbpf submodule
Found clang binary 'clang-14' with version 14 (from 'Ubuntu clang version 14.0.6-++20221029082011+f28c006a5895-1~exp1~20221029202041.162')
libbpf support: custom v0.7.0
  perf_buffer__consume support: no
  btf__load_from_kernel_by_id support: no
  btf__type_cnt support: no
  bpf_object__next_map support: no
  bpf_object__next_program support: no
  bpf_program__insn_cnt support: no
  bpf_map_create support: no
  perf_buffer__new_raw support: no
  bpf_xdp_attach support: no
zlib support: yes
ELF support: yes
pcap support: yes
secure_getenv support: yes
anindya@ubuntu:~/99/bpf-examples$ make

lib

  libbpf
    CC       install/lib/libbpf.a
    INSTALL  install/lib/libbpf.a

  libxdp
    CC       install/lib/libxdp.a
libxdp.c:243:20: error: static declaration of ‘btf__load_from_kernel_by_id’ follows non-static declaration
  243 | static struct btf *btf__load_from_kernel_by_id(__u32 id)
      |                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from libxdp.c:29:
/home/anindya/99/bpf-examples/lib/install/include/bpf/btf.h:121:24: note: previous declaration of ‘btf__load_from_kernel_by_id’ was here
  121 | LIBBPF_API struct btf *btf__load_from_kernel_by_id(__u32 id);
      |                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~
libxdp.c: In function ‘btf__load_from_kernel_by_id’:
libxdp.c:248:2: error: ‘btf__get_from_id’ is deprecated: libbpf v0.6+: use btf__load_from_kernel_by_id instead [-Werror=deprecated-declarations]
  248 |  err = btf__get_from_id(id, &btf);
      |  ^~~
In file included from libxdp.c:29:
/home/anindya/99/bpf-examples/lib/install/include/bpf/btf.h:124:16: note: declared here
  124 | LIBBPF_API int btf__get_from_id(__u32 id, struct btf **btf);
      |                ^~~~~~~~~~~~~~~~
libxdp.c: At top level:
libxdp.c:256:14: error: static declaration of ‘btf__type_cnt’ follows non-static declaration
  256 | static __u32 btf__type_cnt(const struct btf *btf)
      |              ^~~~~~~~~~~~~
In file included from libxdp.c:29:
/home/anindya/99/bpf-examples/lib/install/include/bpf/btf.h:137:18: note: previous declaration of ‘btf__type_cnt’ was here
  137 | LIBBPF_API __u32 btf__type_cnt(const struct btf *btf);
      |                  ^~~~~~~~~~~~~
libxdp.c: In function ‘btf__type_cnt’:
libxdp.c:259:2: error: ‘btf__get_nr_types’ is deprecated: libbpf v0.7+: use btf__type_cnt() instead; note that btf__get_nr_types() == btf__type_cnt() - 1 [-Werror=deprecated-declarations]
  259 |  return btf__get_nr_types(btf) + 1;
      |  ^~~~~~
In file included from libxdp.c:29:
/home/anindya/99/bpf-examples/lib/install/include/bpf/btf.h:136:18: note: declared here
  136 | LIBBPF_API __u32 btf__get_nr_types(const struct btf *btf);
      |                  ^~~~~~~~~~~~~~~~~
libxdp.c: At top level:
libxdp.c:264:24: error: static declaration of ‘bpf_object__next_map’ follows non-static declaration
  264 | static struct bpf_map *bpf_object__next_map(const struct bpf_object *obj,
      |                        ^~~~~~~~~~~~~~~~~~~~
In file included from libxdp.c:28:
/home/anindya/99/bpf-examples/lib/install/include/bpf/libbpf.h:718:1: note: previous declaration of ‘bpf_object__next_map’ was here
  718 | bpf_object__next_map(const struct bpf_object *obj, const struct bpf_map *map);
      | ^~~~~~~~~~~~~~~~~~~~
libxdp.c: In function ‘bpf_object__next_map’:
libxdp.c:267:2: error: ‘bpf_map__next’ is deprecated: libbpf v0.7+: use bpf_object__next_map() instead [-Werror=deprecated-declarations]
  267 |  return bpf_map__next(map, obj);
      |  ^~~~~~
In file included from libxdp.c:28:
/home/anindya/99/bpf-examples/lib/install/include/bpf/libbpf.h:716:17: note: declared here
  716 | struct bpf_map *bpf_map__next(const struct bpf_map *map, const struct bpf_object *obj);
      |                 ^~~~~~~~~~~~~
libxdp.c: At top level:
libxdp.c:272:28: error: static declaration of ‘bpf_object__next_program’ follows non-static declaration
  272 | static struct bpf_program *bpf_object__next_program(const struct bpf_object *obj,
      |                            ^~~~~~~~~~~~~~~~~~~~~~~~
In file included from libxdp.c:28:
/home/anindya/99/bpf-examples/lib/install/include/bpf/libbpf.h:269:1: note: previous declaration of ‘bpf_object__next_program’ was here
  269 | bpf_object__next_program(const struct bpf_object *obj, struct bpf_program *prog);
      | ^~~~~~~~~~~~~~~~~~~~~~~~
libxdp.c: In function ‘bpf_object__next_program’:
libxdp.c:275:2: error: ‘bpf_program__next’ is deprecated: libbpf v0.7+: use bpf_object__next_program() instead [-Werror=deprecated-declarations]
  275 |  return bpf_program__next(prog, obj);
      |  ^~~~~~
In file included from libxdp.c:28:
/home/anindya/99/bpf-examples/lib/install/include/bpf/libbpf.h:266:21: note: declared here
  266 | struct bpf_program *bpf_program__next(struct bpf_program *prog,
      |                     ^~~~~~~~~~~~~~~~~
libxdp.c: At top level:
libxdp.c:281:15: error: static declaration of ‘bpf_program__insn_cnt’ follows non-static declaration
  281 | static size_t bpf_program__insn_cnt(const struct bpf_program *prog)
      |               ^~~~~~~~~~~~~~~~~~~~~
In file included from libxdp.c:28:
/home/anindya/99/bpf-examples/lib/install/include/bpf/libbpf.h:335:19: note: previous declaration of ‘bpf_program__insn_cnt’ was here
  335 | LIBBPF_API size_t bpf_program__insn_cnt(const struct bpf_program *prog);
      |                   ^~~~~~~~~~~~~~~~~~~~~
libxdp.c: In function ‘bpf_program__insn_cnt’:
libxdp.c:285:2: error: ‘bpf_program__size’ is deprecated: libbpf v0.7+: use bpf_program__insn_cnt() instead [-Werror=deprecated-declarations]
  285 |  sz = bpf_program__size(prog);
      |  ^~
In file included from libxdp.c:28:
/home/anindya/99/bpf-examples/lib/install/include/bpf/libbpf.h:301:19: note: declared here
  301 | LIBBPF_API size_t bpf_program__size(const struct bpf_program *prog);
      |                   ^~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
make[4]: *** [Makefile:91: staticobjs/libxdp.o] Error 1
make[3]: *** [Makefile:20: libxdp] Error 2
make[2]: *** [Makefile:28: libxdp] Error 2
make[1]: *** [Makefile:68: install/lib/libxdp.a] Error 2
make: *** [Makefile:34: lib] Error 2
```

**Comment by tohojo** [comment]
It looks like you're not actually using the branch from that pr? Please try switching to the 'configure-fixes' branch.

Also, looks like there's something wrong with the call into the configure of libxdp; could you please try rerunning configure adding `DEBUG_CONFIGURE=1` (along with the clang variables)? - after switching the branch, of course...

**Comment by simosund** [comment]
> Could you please try doing a 'make clean' inside the nat64-bpf
directory, then running 'make' again and posting the full output?

My bad, had a broken bpftool (the one installed by `apt` will error out if you don't use the default kernel). Once I fixed that `nat64-bpf` compiles fine.

Now I instead run into issues with `pkt-loop-filter`.
```
 make
    CLANG    pkt-loop-filter.kern.o
    LLC      pkt-loop-filter.kern.o
    GEN      pkt-loop-filter.kern.skel.h
    CLANG    get-bond-active.kern.o
    LLC      get-bond-active.kern.o
    GEN      get-bond-active.kern.skel.h
    CC       bond-active.o
    CC       pkt-loop-filter
In file included from /usr/include/fcntl.h:314,
                 from pkt-loop-filter.c:7:
In function 'open',
    inlined from 'fork_and_wait' at pkt-loop-filter.c:128:7:
/usr/include/x86_64-linux-gnu/bits/fcntl2.h:50:11: error: call to '__open_missing_mode' declared with attribute error: open with O_CREAT or O_TMPFILE in second argument needs 3 arguments
   50 |           __open_missing_mode ();
      |           ^~~~~~~~~~~~~~~~~~~~~~
make: *** [../lib/common.mk:83: pkt-loop-filter] Error 1

```

**Comment by tohojo** [comment]
@hundredmiles that original issue you were having resurfaced, and should hopefully be fixed properly by #75
